src/app_engine/decline_page.py

```diff
@@ -12,36 +12,99 @@
 import webapp2
 
 import constants
+import gcm_notify
 import gcmrecord
 import room
 import util
 
+
 class DeclinePage(webapp2.RequestHandler):
   def write_response(self, result):
     self.response.write(json.dumps({
-      'result': result
+        'result': result
     }))
-  
+
   def post(self, room_id):
     msg = util.get_message_from_json(self.request.body)
     if not util.has_msg_field(msg, constants.PARAM_CALLEE_GCM_ID, basestring):
       self.write_response(constants.RESPONSE_INVALID_ARGUMENT)
       return
-
     callee_gcm_id = msg[constants.PARAM_CALLEE_GCM_ID]
 
-    # Look up and validate callee by gcm id.
-    # TODO (chuckhays): Once registration is enabled, turn on restriction
+    # Retrieve the room to figure out who the caller is.
+    declined_room = room.get_room(self.request.host_url, room_id)
+
+    # Check that room exists.
+    if declined_room is None:
+      logging.warning(
+          'Client:%s declined unknown room:%s', callee_gcm_id, room_id)
+      self.write_response(constants.RESPONSE_INVALID_ROOM)
+      return
+
+    # Check that the room supports decline.
+    if declined_room.room_type != room.Room.TYPE_DIRECT:
```

Comment:
All the checks should happen inside the cas loop to avoid races.
Why now do remove_room first, if it succeeds, then send GCM, so that you don't need to do checks outside remove_room.